babel/numbers.py

```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
         self.frac_prec = frac_prec
         self.exp_prec = exp_prec
         self.exp_plus = exp_plus
-        if '%' in ''.join(self.prefix + self.suffix):
-            self.scale = 2
-        elif u'‰' in ''.join(self.prefix + self.suffix):
-            self.scale = 3
-        else:
-            self.scale = 0
 
     def __repr__(self):
         return '<%s %r>' % (type(self).__name__, self.pattern)
 
-    def apply(self, value, locale, currency=None, force_frac=None):
-        frac_prec = force_frac or self.frac_prec
+    @property
+    def scale(self):
```

Comment:
This could use the [`@cached_property` decorator](https://github.com/bottlepy/bottle/commit/fa7733e075da0d790d809aa3d2f53071897e6f76#diff-ad8cb2f640fd3a70db3fc97f3044a4e6R147).